src/vm/jithelpers.cpp

```diff
@@ -2215,66 +2141,87 @@ BOOL ObjIsInstanceOf(Object *pObject, TypeHandle toTypeHnd, BOOL throwCastExcept
     } CONTRACTL_END;
 
     BOOL fCast = FALSE;
+    MethodTable* pMT = pObject->GetMethodTable();
 
     OBJECTREF obj = ObjectToOBJECTREF(pObject);
-
     GCPROTECT_BEGIN(obj);
 
-    TypeHandle fromTypeHnd = obj->GetTypeHandle();
-
-    // If we are trying to cast a proxy we need to delegate to remoting
-    // services which will determine whether the proxy and the type are compatible.
-    // Start by doing a quick static cast check to see if the type information captured in
-    // the metadata indicates that the cast is legal.
-    if (fromTypeHnd.CanCastTo(toTypeHnd))
+    if (pMT->IsArray())
     {
-        fCast = TRUE;
+        if (toTypeHnd.IsArray())
+        {
+            fCast = pMT->ArrayIsInstanceOf(toTypeHnd, /* pVisited */ NULL);
+        }
+        else
+        {
+            MethodTable* toMT = toTypeHnd.GetMethodTable();
+            if (toMT->IsInterface() && toMT->HasInstantiation())
+            {
+                fCast = pMT->ArraySupportsBizarreInterface(toMT, /* pVisited */ NULL);
+            }
+            else
+            {
+                fCast = pMT->CanCastToClassOrInterface(toMT, /* pVisited */ NULL);
+            }
+        }
     }
-    else
-#ifdef FEATURE_COMINTEROP
-    // If we are casting a COM object from interface then we need to do a check to see 
-    // if it implements the interface.
-    if (toTypeHnd.IsInterface() && fromTypeHnd.GetMethodTable()->IsComObjectType())
+    else if (toTypeHnd.IsTypeDesc())
     {
-        fCast = ComObject::SupportsInterface(obj, toTypeHnd.AsMethodTable());
+        CastCache::TryAddToCache(pMT, toTypeHnd, FALSE);
+        fCast = FALSE;
     }
-    else
-#endif // FEATURE_COMINTEROP
-    if (Nullable::IsNullableForType(toTypeHnd, obj->GetMethodTable()))
+    else if (pMT->CanCastToClassOrInterface(toTypeHnd.AsMethodTable(), /* pVisited */ NULL))
     {
-        // allow an object of type T to be cast to Nullable<T> (they have the same representation)
         fCast = TRUE;
     }
```

Comment:
Why does this case not add to the cache?

Reply:
CanCastToClassOrInterface should add to the cache, if the result is cacheable. Ex: if it was a ComClass or ICastable, it may be not cacheable. Caller does not always know that part.
 
I generally try checking the cache closer to the entry points (so that I could return result early), but add new result to the cache at the leaf methods - where the decision is made. 
That is not always possible since interaction between helpers could be complex, but that is the general idea.

---
In reply to: [306594926](https://github.com/dotnet/coreclr/pull/23548#discussion_r306594926) [](ancestors = 306594926)